### ISSUE TITLE
Cryptography 41.0.2 fails to parse certificate signed with DSA

### ISSUE THREAD

**Issue by niooss-ledger** [pyca/cryptography#9253]
Hello,

When parsing a certificated signed with DSA, `cryptography` 41.0.2 fails with:

```text
  File "/usr/local/lib/python3.11/site-packages/cryptography/x509/base.py", line 583, in load_pem_x509_certificate
    return rust_x509.load_pem_x509_certificate(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: error parsing asn1 value: ParseError { kind: ExtraData, location: ["Certificate::tbs_cert", "TbsCertificate::signature_alg"] }
```

To reproduce the bug, I generated a certificate in a clean container `python:3.11-alpine` container (launched with `docker run --rm -it docker.io/library/python:3.11-alpine sh`):

```sh
apk add openjdk11 openssl
keytool -genkeypair -noprompt -keyalg dsa -storetype pkcs12 -keystore store.p12 -storepass changeit -alias a -dname CN=example
openssl pkcs12 -info -passin pass:changeit -nodes -in store.p12
```

This generated a DSA keypair and displayed a certificate.
I load this certificate with `cryptography.x509.load_pem_x509_certificate`:

```python
import cryptography.hazmat.backends
import cryptography.x509

certificate = b"""-----BEGIN CERTIFICATE-----
MIIEIDCCA8ugAwIBAgIEK3rzQDANBglghkgBZQMEAwIFADASMRAwDgYDVQQDEwdl
eGFtcGxlMB4XDTIzMDcxNzEzMDQ0MVoXDTIzMTAxNTEzMDQ0MVowEjEQMA4GA1UE
AxMHZXhhbXBsZTCCA0IwggI1BgcqhkjOOAQBMIICKAKCAQEAj3k12bmq6b+r7Yh6
z0lRtvMuxZ47rzcY6OrElh8+/TYG50NRqcQYMzm4CefCrhxTm6dHW4XQEa24tHmH
dUmEaVysDo8UszYIKKIv+icRCj1iqZNFNAmg/mlsRlj4S90ggZw3CaAQV7GVrc0A
Iz26VIS2KR+dZI74g0SGd5ec7AS0NKasLnXpmF3iPbApL8ERjJ/6nYGB5zONt5K3
MNe540lZL2gJmHIVORXqPWuLRlPGM0WPgDsypMLg8nKQJW5OP4o7CDihxFDk4Ywa
KaN9316hQ95LZv8EkD7VzxYj4VjUh8YI6X8hHNgdyiPLbjgHZfgi40K+SEwFdjk5
YBzWZwIdALr2lqaFePff3uf6Z8l3x4XvMrIzuuWAwLzVaV0CggEAFqZcWCBIUHBO
dQKjl1cEDTTaOjR4wVTU5KXALSQu4E+W5h5L0JBKvayPN+6x4J8xgtI8kEPLZC+I
AEFg7fnKCbMgdqecMqYn8kc+kYebosTnRL0ggVRMtVuALDaNH6g+1InpTg+gaI4y
QopceMR4xo0FJ7ccmjq7CwvhLERoljnn08502xAaZaorh/ZMaCbbPscvS1WZg0u0
7bAvfJDppJbTpV1TW+v8RdT2GfY/Pe27hzklwvIk4HcxKW2oh+weR0j4fvtf3rdU
hDFrIjLe5VPdrwIRKw0fAtowlzIk/ieu2oudSyki2bqL457Z4QOmPFKBC8aIt+Lt
Qxbh7xfb3gOCAQUAAoIBAE9198Qk2RSPzrlp1Kn2dekLQl/7vhwIMM2ELDO16y3F
pUI+m3NRUDThVos5La8qcnJpBVG/uM3nkpSev10zAUEv7DZoXWqmM71JXyk0dJrv
ZHJMnsLAC9raNxQNPLeXBowprVXCOFwLZHplYNDi6kq2/scg87YC6l3UYjaVa4gD
jIDYmKcKuVI61JIYDJUIBfDtB/bgvAPwOf7Nb++lNQ4X5ibBrnuIGA5IXghxs3+3
NS1xoDUMt3bArjex2q3L/s0+ag5x2VbxqtQhMfBo9Tz5/R2vbN3NUn3oqcS1nBPx
fR7YyV+6Wvc7Lqixx+3KWPuzsB/xbkUQb4jtKCJ4GYijITAfMB0GA1UdDgQWBBQ2
zCtCCA1f4l/9+S+kzPnCD0q1UDANBglghkgBZQMEAwIFAANAADA9Ah0AkTHX6Lxh
sobhhLgQWHiHOZ8ndECVDfG5ro3NtQIcEEkRmOSDscN78wHoMqSoSl+v2kAFWXKN
SYhhaQ==
-----END CERTIFICATE-----"""
backend = cryptography.hazmat.backends.default_backend()
cert = cryptography.x509.load_pem_x509_certificate(certificate, backend)
print(cert)
```

With `cryptography==41.0.2`, this fails with

```text
Traceback (most recent call last):
  File "/example.py", line 30, in <module>
    cert = cryptography.x509.load_pem_x509_certificate(certificate, backend)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/cryptography/x509/base.py", line 583, in load_pem_x509_certificate
    return rust_x509.load_pem_x509_certificate(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: error parsing asn1 value: ParseError { kind: ExtraData, location: ["Certificate::tbs_cert", "TbsCertificate::signature_alg"] }
```

- This is also broken with `cryptography==41.0.0`
- This works when downgrading to `cryptography==40.0.2` and previous versions (the Python script successfully displays `<Certificate(subject=<Name(CN=example)>, ...)>`).
- This also fails with current git `main` installed with `apk add cargo git pkgconf openssl-dev && pip install git+https://github.com/pyca/cryptography`.

I am using Python 3.11, `cryptography` 41.0.2, `cffi` 1.15.1 and `pip` 23.1.2, on Linux (Alpine container) on an x86-64 processor.

Are certificates signed with DSA unsupported since `cryptography` 41.0.0?

**Comment by niooss-ledger** [comment]
The bug still exists on OpenJDK 17 and 20.

To reproduce it:
- I used an Arch Linux container: `docker run --rm -it docker.io/library/archlinux bash`
- I installed Python and OpenJDK: `pacman -Sy python-pip jdk-openjdk`
- The version of OpenJDK which was installed was `jdk-openjdk 20.0.1.u9-3`
- I generated a PKCS#12 keystore with `/usr/lib/jvm/java-20-openjdk/bin/keytool -genkeypair -noprompt -keyalg dsa -storetype pkcs12 -keystore store.p12 -storepass changeit -alias a -dname CN=example`
- I converted the certificate in the keystore to PEM with `openssl pkcs12 -info -passin pass:changeit -nodes -in store.p12`

The certificate was:

```text
-----BEGIN CERTIFICATE-----
MIIEJDCCA9CgAwIBAgIJANXeXtYKjIwEMA0GCWCGSAFlAwQDAgUAMBIxEDAOBgNV
BAMTB2V4YW1wbGUwHhcNMjMwNzE4MTQ0OTEwWhcNMjMxMDE2MTQ0OTEwWjASMRAw
DgYDVQQDEwdleGFtcGxlMIIDQjCCAjUGByqGSM44BAEwggIoAoIBAQCPeTXZuarp
v6vtiHrPSVG28y7FnjuvNxjo6sSWHz79NgbnQ1GpxBgzObgJ58KuHFObp0dbhdAR
rbi0eYd1SYRpXKwOjxSzNggooi/6JxEKPWKpk0U0CaD+aWxGWPhL3SCBnDcJoBBX
sZWtzQAjPbpUhLYpH51kjviDRIZ3l5zsBLQ0pqwudemYXeI9sCkvwRGMn/qdgYHn
M423krcw17njSVkvaAmYchU5Feo9a4tGU8YzRY+AOzKkwuDycpAlbk4/ijsIOKHE
UOThjBopo33fXqFD3ktm/wSQPtXPFiPhWNSHxgjpfyEc2B3KI8tuOAdl+CLjQr5I
TAV2OTlgHNZnAh0AuvaWpoV499/e5/pnyXfHhe8ysjO65YDAvNVpXQKCAQAWplxY
IEhQcE51AqOXVwQNNNo6NHjBVNTkpcAtJC7gT5bmHkvQkEq9rI837rHgnzGC0jyQ
Q8tkL4gAQWDt+coJsyB2p5wypifyRz6Rh5uixOdEvSCBVEy1W4AsNo0fqD7UielO
D6BojjJCilx4xHjGjQUntxyaOrsLC+EsRGiWOefTznTbEBplqiuH9kxoJts+xy9L
VZmDS7TtsC98kOmkltOlXVNb6/xF1PYZ9j897buHOSXC8iTgdzEpbaiH7B5HSPh+
+1/et1SEMWsiMt7lU92vAhErDR8C2jCXMiT+J67ai51LKSLZuovjntnhA6Y8UoEL
xoi34u1DFuHvF9veA4IBBQACggEASGXf9G1yAHOU7G/su00m3SricigX9zPfUh89
sl9lj5Ht6c545WGSkg2vjPbK4KFfDeWTzz2neKoM9xWxJUjGBfURX3+b6BpAj86x
+vQok87c16mpw8Cf3MtAgc0oHOM1I7pGDO/9/ZvKbSMMB0S7Sv5Q99VYWGN2od+m
yVzz/oYNy9IRSFfNaHPOweMP7oQFYNNfoc9jXCXayICoj9IDFDSvvtC71wc8Z7Ey
b2VAUUMwpSQ+3nrurTCfoCkE9MQSC01GICeIK8Es+EUuV8jUfV+XAmBpul8jwsxi
+XcM4arXoru/9V8qaS+qGsz1dCqAp4hcqq3D20OkJvI3JMu316MhMB8wHQYDVR0O
BBYEFMaC7UTR+brmFDATYd5Oss2eshvCMA0GCWCGSAFlAwQDAgUAAz8AMDwCHFqb
CNDLKq90HzhWw88QgwcwlGnu6W5RCeS8PCICHGqdeHyfftezEfGWp3qIpzlUa2oG
GizOpATUqSQ=
-----END CERTIFICATE-----
```

In `openssl asn1parse -inform PEM -i` the ASN.1 structure starts with:

```text
    0:d=0  hl=4 l=1060 cons: SEQUENCE          
    4:d=1  hl=4 l= 976 cons:  SEQUENCE          
    8:d=2  hl=2 l=   3 cons:   cont [ 0 ]        
   10:d=3  hl=2 l=   1 prim:    INTEGER           :02
   13:d=2  hl=2 l=   9 prim:   INTEGER           :D5DE5ED60A8C8C04
   24:d=2  hl=2 l=  13 cons:   SEQUENCE          
   26:d=3  hl=2 l=   9 prim:    OBJECT            :dsa_with_SHA256
   37:d=3  hl=2 l=   0 prim:    NULL              
   39:d=2  hl=2 l=  18 cons:   SEQUENCE 
```

So a `NULL` value was added after the OID `dsa_with_SHA256`. And `cryptography` fails:

```console
$ python3 -m venv venv
$ ./venv/bin/pip install cryptography
Collecting cryptography
  Downloading cryptography-41.0.2-cp37-abi3-manylinux_2_28_x86_64.whl (4.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 27.1 MB/s eta 0:00:00
Collecting cffi>=1.12
  Downloading cffi-1.15.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (462 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 462.6/462.6 kB 50.2 MB/s eta 0:00:00
Collecting pycparser
  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.7/118.7 kB 75.0 MB/s eta 0:00:00
Installing collected packages: pycparser, cffi, cryptography
Successfully installed cffi-1.15.1 cryptography-41.0.2 pycparser-2.21

[notice] A new release of pip available: 22.3.1 -> 23.2
[notice] To update, run: python3 -m pip install --upgrade pip
$ ./venv/bin/python3
Python 3.11.3 (main, Jun  5 2023, 09:32:32) [GCC 13.1.1 20230429] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cryptography.hazmat.backends
>>> import cryptography.x509
>>> certificate = b"""-----BEGIN CERTIFICATE-----
... MIIEJDCCA9CgAwIBAgIJANXeXtYKjIwEMA0GCWCGSAFlAwQDAgUAMBIxEDAOBgNV
... BAMTB2V4YW1wbGUwHhcNMjMwNzE4MTQ0OTEwWhcNMjMxMDE2MTQ0OTEwWjASMRAw
... DgYDVQQDEwdleGFtcGxlMIIDQjCCAjUGByqGSM44BAEwggIoAoIBAQCPeTXZuarp
... v6vtiHrPSVG28y7FnjuvNxjo6sSWHz79NgbnQ1GpxBgzObgJ58KuHFObp0dbhdAR
... rbi0eYd1SYRpXKwOjxSzNggooi/6JxEKPWKpk0U0CaD+aWxGWPhL3SCBnDcJoBBX
... sZWtzQAjPbpUhLYpH51kjviDRIZ3l5zsBLQ0pqwudemYXeI9sCkvwRGMn/qdgYHn
... M423krcw17njSVkvaAmYchU5Feo9a4tGU8YzRY+AOzKkwuDycpAlbk4/ijsIOKHE
... UOThjBopo33fXqFD3ktm/wSQPtXPFiPhWNSHxgjpfyEc2B3KI8tuOAdl+CLjQr5I
... TAV2OTlgHNZnAh0AuvaWpoV499/e5/pnyXfHhe8ysjO65YDAvNVpXQKCAQAWplxY
... IEhQcE51AqOXVwQNNNo6NHjBVNTkpcAtJC7gT5bmHkvQkEq9rI837rHgnzGC0jyQ
... Q8tkL4gAQWDt+coJsyB2p5wypifyRz6Rh5uixOdEvSCBVEy1W4AsNo0fqD7UielO
... D6BojjJCilx4xHjGjQUntxyaOrsLC+EsRGiWOefTznTbEBplqiuH9kxoJts+xy9L
... VZmDS7TtsC98kOmkltOlXVNb6/xF1PYZ9j897buHOSXC8iTgdzEpbaiH7B5HSPh+
... +1/et1SEMWsiMt7lU92vAhErDR8C2jCXMiT+J67ai51LKSLZuovjntnhA6Y8UoEL
... xoi34u1DFuHvF9veA4IBBQACggEASGXf9G1yAHOU7G/su00m3SricigX9zPfUh89
... sl9lj5Ht6c545WGSkg2vjPbK4KFfDeWTzz2neKoM9xWxJUjGBfURX3+b6BpAj86x
... +vQok87c16mpw8Cf3MtAgc0oHOM1I7pGDO/9/ZvKbSMMB0S7Sv5Q99VYWGN2od+m
... yVzz/oYNy9IRSFfNaHPOweMP7oQFYNNfoc9jXCXayICoj9IDFDSvvtC71wc8Z7Ey
... b2VAUUMwpSQ+3nrurTCfoCkE9MQSC01GICeIK8Es+EUuV8jUfV+XAmBpul8jwsxi
... +XcM4arXoru/9V8qaS+qGsz1dCqAp4hcqq3D20OkJvI3JMu316MhMB8wHQYDVR0O
... BBYEFMaC7UTR+brmFDATYd5Oss2eshvCMA0GCWCGSAFlAwQDAgUAAz8AMDwCHFqb
... CNDLKq90HzhWw88QgwcwlGnu6W5RCeS8PCICHGqdeHyfftezEfGWp3qIpzlUa2oG
... GizOpATUqSQ=
... -----END CERTIFICATE-----
... """
>>> backend = cryptography.hazmat.backends.default_backend()
>>> cert = cryptography.x509.load_pem_x509_certificate(certificate, backend)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "//venv/lib/python3.11/site-packages/cryptography/x509/base.py", line 583, in load_pem_x509_certificate
    return rust_x509.load_pem_x509_certificate(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: error parsing asn1 value: ParseError { kind: ExtraData, location: ["Certificate::tbs_cert", "TbsCertificate::signature_alg"] }
```

**Comment by seanjmullan** [comment]
This is an issue that has been fixed in JDK 21, targeted to be released in September: https://bugs.openjdk.org/browse/JDK-8301793

We can consider backporting the fix to earlier JDK versions, but OTOH I'm a little wary of changing an encoding in previous versions that has worked w/o issue for many years. Some more bake time in JDK 21 is probably warranted before considering that.
